README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![Bolder Flight Systems Logo](img/logo-words_75.png) &nbsp; &nbsp; ![Arduino Logo](img/arduino_logo_75.png)

# Lis3mdl
This library communicates with the [LIS3MDL](https://www.st.com/en/mems-and-sensors/lis3mdl.html) magnetometer. This library is compatible with Arduino and CMake build systems.
   * [License](LICENSE.md)
   * [Changelog](CHANGELOG.md)
   * [Contributing guide](CONTRIBUTING.md)

# Description
The Invense LIS3MDL is a three-axis magnetometer, which supports I2C, up to 400 kHz, and SPI communication, up to 10 MHz. The following selectable full scale sensor ranges and output data rates are available:

| Full Scale Range |
| --- |
| +/- 4 Gauss |
| +/- 8 Gauss |
| +/- 12 Gauss |
| +/- 16 Gauss |

| Output Data Rate |
| --- |
| 0.625HZ |
| 1.25HZ |
| 2.5HZ |
| 5HZ |
| 10HZ |
| 20HZ |
| 40HZ |
| 80HZ |
| 155HZ |
| 300HZ |
| 560HZ |
| 1000HZ |

# Installation

## Arduino
Use the Arduino Library Manager to install this library or clone to your Arduino/libraries folder. This library is added as:

```C++
#include "lis3mdl.h"
```

Example Arduino executables are located in: *examples/arduino/*. Teensy 3.x, 4.x, and LC devices are used for testing under Arduino and this library should be compatible with other Arduino devices.

## CMake
CMake is used to build this library, which is exported as a library target called *lis3mdl*. This library is added as:

```C++
#include "lis3mdl.h"
```

The library can be also be compiled stand-alone using the CMake idiom of creating a *build* directory and then, from within that directory issuing:

```
cmake .. -DMCU=MK66FX1M0
make
```

This will build the library and example executables called *i2c_example* and *spi_example*. The example executable source files are located at *examples/cmake*. Notice that the *cmake* command includes a define specifying the microcontroller the code is being compiled for. This is required to correctly configure the code, CPU frequency, and compile/linker options. The available MCUs are:
   * MK20DX128
   * MK20DX256
   * MK64FX512
   * MK66FX1M0
   * MKL26Z64
   * IMXRT1062_T40
   * IMXRT1062_T41
   * IMXRT1062_MMOD

These are known to work with the same packages used in Teensy products. Also switching packages is known to work well, as long as it's only a package change.

The example targets create executables for communicating with the sensor using I2C or SPI communication, using the data ready interrupt, and using the wake on motion interrupt, respectively. Each target also has a *_hex*, for creating the hex file to upload to the microcontroller, and an *_upload* for using the [Teensy CLI Uploader](https://www.pjrc.com/teensy/loader_cli.html) to flash the Teensy. Please note that instructions for setting up your build environment can be found in our [build-tools repo](https://github.com/bolderflight/build-tools).

# Namespace
This library is within the namespace *bfs*.

# Lis3mdl

## Methods

**Lis3mdl()** Default constructor, requires calling the Config method to setup the I2C or SPI bus and I2C address or SPI chip select pin.

**Lis3mdl(TwoWire &ast;i2c, const I2cAddr addr)** Creates a *Lis3mdl* object using I2C communication. A pointer to the bus is passed along with the I2C address. The primary address (0x1C) should be used if SDO is grounded and the secondary address (0x1E) if SDO is pulled high.

```C++
bfs::Lis3mdl mag(&Wire, bfs::Lis3mdl::I2C_ADDR_PRIM);
```

**Lis3mdl(SPIClass &ast;spi, const uint8_t cs)** Creates a *Lis3mdl* object using SPI communication. A pointer to the bus is passed along with the chip select pin number. Any digital pin can be used as chip select.

```C++
bfs::Lis3mdl mag(&SPI, 10);
```

**void Config(TwoWire &ast;bus, const I2cAddr addr)** This is required when using the default constructor and sets up the I2C bus and I2C address.

**void Config(SPIClass &ast;spi, const uint8_t cs)** This is required when using the default constructor and sets up the SPI bus and chip select pin.

**bool Begin()** Initializes communication with the sensor and sets a default configuration. True is returned on success and false on failure. Note that the bus is not initialized in this method to improve compatibility with other sensors on the bus. The default configuration is:

| Range | Output Data Rate |
| --- | --- |
| +/-16 Gauss | 155 Hz |

```C++
Wire.begin();
Wire.setClock(400000);
if (!mag.Begin()) {
  // error handling
}
```

**bool ConfigOdr(const Odr odr)** Configures the output data rate. True is returned on success and false on failure. The available options are:

| Enum | Rate |
| --- | --- |
| ODR_0_625HZ | 0.625Hz |
| ODR_1_25HZ | 1.25Hz |
| ODR_2_5HZ | 2.5Hz |
| ODR_5HZ | 5HZ |
| ODR_10HZ | 10HZ |
| ODR_20HZ | 20HZ |
| ODR_40HZ | 40HZ |
| ODR_80HZ | 80HZ |
| ODR_155HZ | 155HZ |
| ODR_300HZ | 300HZ |
| ODR_560HZ | 560HZ |
| ODR_1000HZ | 1000HZ |

```C++
if(!mag.ConfigOdr(bfs::Lis3mdl::ODR_10HZ)) {
  // error handling
}
```

**Odr odr()** Returns the current output data rate.

**bool ConfigRange(const Range range)** Configures the sensor full scale range. True is returned on success and false on failure. The available options are:

| Enum | Range |
| --- | --- |
| RANGE_4GS | +/- 4 Gauss |
| RANGE_8GS | +/- 8 Gauss |
| RANGE_12GS | +/- 12 Gauss |
| RANGE_16GS | +/- 16 Gauss |

```C++
if(!mag.ConfigRange(bfs::Lis3mdl::RANGE_12GS)) {
  // error handling
}
```

**Range range()** Returns the current full scale range.

**bool Read()** Reads data from the sensor and stores the data in the *Lis3mdl* object. Note that new data from each axis is checked individually. If data from any axis is updated, this method returns true. When using the data ready interrupt to call this method, all three axes are typically updated simultaneously. When polling this method, typically x and y are updated each frame at the specified output data rate. The z axis data sometimes misses a frame.

```C++
if (mag.Read()) {
  // data processing
}
```

**bool new_x_data()** Returns true if the x axis data has been updated. This flag is latched and reset when this method is called. Similar methods exist for y and z axis.

```C++
bool new_data[3];
new_data[0] = mag.new_x_data();
new_data[1] = mag.new_y_data();
new_data[2] = mag.new_z_data();
```

**float mag_x_ut()** Returns the most recent x axis measurement from the *Lis3mdl* object in uT. Similar methods exist for y and z axis.

```C++
float data[3];
data[0] = mag.mag_x_ut();
data[1] = mag.mag_y_ut();
data[2] = mag.mag_z_ut();
```

**float die_temp_c()** Returns the most recent temperature measurement from the *Lis3mdl* object in degrees C.


CHANGELOG.md
# Changelog

## v1.2.0
- Added die temperature data

## v1.1.5
- Updated core to v3.1.3

## v1.1.4
- Fixed typo in SPI mode

## v1.1.3
- Updated core to v3.1.2

## v1.1.2
- Changed SPI to MODE3 for better compatibility with other sensors

## v1.1.1
- Updated core to support MMOD

## v1.1.0
- Added default constructor

## v1.0.0
- Initial baseline


src/lis3mdl.cpp
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2022 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "lis3mdl.h"  // NOLINT

namespace bfs {

void Lis3mdl::Config(TwoWire *i2c, const I2cAddr addr) {
  i2c_ = i2c;
  dev_ = static_cast<uint8_t>(addr);
  iface_ = I2C;
}

void Lis3mdl::Config(SPIClass *spi, const uint8_t cs) {
  spi_ = spi;
  dev_ = cs;
  iface_ = SPI;
}

bool Lis3mdl::Begin() {
  if (iface_ == SPI) {
    pinMode(dev_, OUTPUT);
    /* Toggle CS pin to lock in SPI mode */
    digitalWrite(dev_, LOW);
    delay(1);
    digitalWrite(dev_, HIGH);
    delay(1);
  }
  /* Check WHOAMI */
  if (!ReadRegisters(REG_WHO_AM_I_, sizeof(who_am_i_), &who_am_i_)) {
    return false;
  }
  if (who_am_i_ != LIS3MDL_WHOAMI_) {
    return false;
  }
  /* Enable temp sensor */
  if (!EnableTemp()) {
    return false;
  }
  /* Set range to +/-16GS */
  if (!ConfigRange(RANGE_16GS)) {
    return false;
  }
  /* Set ODR to 155 Hz */
  if (!ConfigOdr(ODR_155HZ)) {
    return false;
  }
  /* Continuous mode */
  if (!EnableContinousMode(true)) {
    return false;
  }
  /* Enable Block Update */
  if (!EnableBdu()) {
    return false;
  }
  while (!Read()) {}
  return true;
}
bool Lis3mdl::ConfigOdr(const Odr odr) {
  if (odr == ODR_155HZ) {
    if (!SetPerfMode(PERF_MODE_ULTRA_HIGH)) {
      return false;
    }
    if (!GetPerfMode(&perf_mode_)) {
      return false;
    }
    if (perf_mode_ != PERF_MODE_ULTRA_HIGH) {
      return false;
    }
    continuous_ = true;
  } else if (odr == ODR_300HZ) {
    if (!SetPerfMode(PERF_MODE_HIGH)) {
      return false;
    }
    if (!GetPerfMode(&perf_mode_)) {
      return false;
    }
    if (perf_mode_ != PERF_MODE_HIGH) {
      return false;
    }
    continuous_ = true;
  } else if (odr == ODR_560HZ) {
    if (!SetPerfMode(PERF_MODE_MEDIUM_POWER)) {
      return false;
    }
    if (!GetPerfMode(&perf_mode_)) {
      return false;
    }
    if (perf_mode_ != PERF_MODE_MEDIUM_POWER) {
      return false;
    }
    continuous_ = true;
  } else if (odr == ODR_1000HZ) {
    if (!SetPerfMode(PERF_MODE_LOW_POWER)) {
      return false;
    }
    if (!GetPerfMode(&perf_mode_)) {
      return false;
    }
    if (perf_mode_ != PERF_MODE_LOW_POWER) {
      return false;
    }
    continuous_ = true;
  } else if (odr == ODR_80HZ) {
    continuous_ = true;
  } else {
    continuous_ = false;
  }
  delay(50);
  if (!SetDataRate(odr)) {
    return false;
  }
  if (!GetDataRate(&odr_)) {
    return false;
  }
  if (odr != odr_) {
    return false;
  }
  if (!EnableContinousMode(continuous_)) {
    return false;
  }
  return true;
}
bool Lis3mdl::ConfigRange(const Range range) {
  if (!SetRange(range)) {
    return false;
  }
  if (!GetRange(&range_)) {
    return false;
  }
  if (range_ != range) {
    return false;
  }
  switch (range) {
    case RANGE_4GS: {
      scale_ = 6842.0f;
      break;
    }
    case RANGE_8GS: {
      scale_ = 3421.0f;
      break;
    }
    case RANGE_12GS: {
      scale_ = 2281.0f;
      break;
    }
    case RANGE_16GS: {
      scale_ = 1711.0f;
      break;
    }
  }
  return true;
}
bool Lis3mdl::Read() {
  if (!ReadRegisters(REG_STATUS_REG_, 1, &stat_)) {
    return false;
  }
  data_updated_ = false;
  if (stat_ & 0x01) {
    data_updated_ = true;
    new_x_data_ = true;
  }
  if (stat_ & 0x02) {
    data_updated_ = true;
    new_y_data_ = true;
  }
  if (stat_ & 0x04) {
    data_updated_ = true;
    new_z_data_ = true;
  }
  if (new_x_data_) {
    if (ReadRegisters(REG_OUT_X_L_, 2, buf_)) {
      x_ = static_cast<int16_t>(buf_[1]) << 8 | buf_[0];
      mag_ut_[0] = static_cast<float>(x_) / scale_ * 100.0f;
    }
  }
  if (new_y_data_) {
    if (ReadRegisters(REG_OUT_Y_L_, 2, buf_)) {
      y_ = static_cast<int16_t>(buf_[1]) << 8 | buf_[0];
      mag_ut_[1] = static_cast<float>(y_) / scale_ * 100.0f;
    }
  }
  if (new_z_data_) {
    if (ReadRegisters(REG_OUT_Z_L_, 2, buf_)) {
      z_ = static_cast<int16_t>(buf_[1]) << 8 | buf_[0];
      mag_ut_[2] = static_cast<float>(z_) / scale_ * 100.0f;
    }
  }
  if (ReadRegisters(REG_TEMP_OUT_L_, 2, buf_)) {
    t_ = static_cast<int16_t>(buf_[1]) << 8 | buf_[0];
    temp_ = static_cast<float>(t_) / 8.0f + 25.0f;
  }
  return data_updated_;
}
bool Lis3mdl::new_x_data() {
  if (new_x_data_) {
    new_x_data_ = false;
    return true;
  }
  return false;
}
bool Lis3mdl::new_y_data() {
  if (new_y_data_) {
    new_y_data_ = false;
    return true;
  }
  return false;
}
bool Lis3mdl::new_z_data() {
  if (new_z_data_) {
    new_z_data_ = false;
    return true;
  }
  return false;
}
bool Lis3mdl::EnableBdu() {
  if (!Write(REG_CTRL_REG5_, 0x01, 1, 6)) {
    return false;
  }
  return true;
}
bool Lis3mdl::EnableContinousMode(bool en) {
  if (en) {
    if (!SetOpMode(OP_MODE_CONTINUOUS)) {
      return false;
    }
    if (!GetOpMode(&op_mode_)) {
      return false;
    }
    if (op_mode_ != OP_MODE_CONTINUOUS) {
      return false;
    }
    return true;
  } else {
    if (!SetOpMode(OP_MODE_SINGLE)) {
      return false;
    }
    if (!GetOpMode(&op_mode_)) {
      return false;
    }
    if (op_mode_ != OP_MODE_SINGLE) {
      return false;
    }
    return true;
  }
}
bool Lis3mdl::EnableTemp() {
  if (!Write(REG_CTRL_REG1_, 0x01, 1, 7)) {
    return false;
  }
  return true;
}
bool Lis3mdl::SetPerfMode(const PerfMode perf_mode) {
  if (!Write(REG_CTRL_REG1_, static_cast<uint8_t>(perf_mode), 2, 5)) {
    return false;
  }
  if (!Write(REG_CTRL_REG4_, static_cast<uint8_t>(perf_mode), 2, 2)) {
    return false;
  }
  return true;
}
bool Lis3mdl::SetOpMode(const OpMode op_mode) {
  if (!Write(REG_CTRL_REG3_, static_cast<uint8_t>(op_mode), 2, 0)) {
    return false;
  }
  return true;
}
bool Lis3mdl::SetDataRate(const Odr data_rate) {
  if (!Write(REG_CTRL_REG1_, static_cast<uint8_t>(data_rate), 4, 1)) {
    return false;
  }
  return true;
}
bool Lis3mdl::SetRange(const Range range) {
  if (!Write(REG_CTRL_REG2_, static_cast<uint8_t>(range), 2, 5)) {
    return false;
  }
  return true;
}
bool Lis3mdl::GetPerfMode(PerfMode * const perf_mode) {
  if (!Read(REG_CTRL_REG1_, 2, 5, &buf_[0])) {
    return false;
  }
  if (!Read(REG_CTRL_REG4_, 2, 2, &buf_[1])) {
    return false;
  }
  if (buf_[0] != buf_[1]) {
    return false;
  }
  (*perf_mode) = static_cast<PerfMode>(buf_[0]);
  return true;
}
bool Lis3mdl::GetOpMode(OpMode * const op_mode) {
  if (!Read(REG_CTRL_REG3_, 2, 0, &buf_[0])) {
    return false;
  }
  (*op_mode) = static_cast<OpMode>(buf_[0]);
  return true;
}
bool Lis3mdl::GetDataRate(Odr * const data_rate) {
  if (!Read(REG_CTRL_REG1_, 4, 1, &buf_[0])) {
    return false;
  }
  (*data_rate) = static_cast<Odr>(buf_[0]);
  return true;
}
bool Lis3mdl::GetRange(Range * const range) {
  if (!Read(REG_CTRL_REG2_, 2, 5, &buf_[0])) {
    return false;
  }
  (*range) = static_cast<Range>(buf_[0]);
  return true;
}
bool Lis3mdl::Read(const uint8_t reg, const uint8_t bits, const uint8_t shift,
                   uint8_t * const data) {
  if (!ReadRegisters(reg, 1, &val_)) {
    return false;
  }
  val_ >>= shift;
  mask_ = (1 << (bits)) - 1;
  *data = val_ & mask_;
  return true;
}
bool Lis3mdl::Write(const uint8_t reg, const uint8_t data, const uint8_t bits,
                    const uint8_t shift) {
  if (!ReadRegisters(reg, 1, &val_)) {
    return false;
  }
  mask_ = (1 << (bits)) - 1;
  data_ = data & mask_;
  mask_ <<= shift;
  val_ &= ~mask_;
  val_ |= data_ << shift;
  return WriteRegister(reg, val_);
}
bool Lis3mdl::WriteRegister(const uint8_t reg, const uint8_t data) {
  uint8_t ret_val;
  if (iface_ == I2C) {
    i2c_->beginTransmission(dev_);
    i2c_->write(reg);
    i2c_->write(data);
    i2c_->endTransmission();
  } else {
    spi_->beginTransaction(SPISettings(SPI_CLOCK_, MSBFIRST, SPI_MODE3));
    #if defined(TEENSYDUINO)
    digitalWriteFast(dev_, LOW);
    #else
    digitalWrite(dev_, LOW);
    #endif
    spi_->transfer(reg);
    spi_->transfer(data);
    #if defined(TEENSYDUINO)
    digitalWriteFast(dev_, HIGH);
    #else
    digitalWrite(dev_, HIGH);
    #endif
    spi_->endTransaction();
  }
  delay(10);
  ReadRegisters(reg, sizeof(ret_val), &ret_val);
  if (data == ret_val) {
    return true;
  } else {
    return false;
  }
}
bool Lis3mdl::ReadRegisters(const uint8_t reg, const uint8_t count,
                            uint8_t * const data) {
  if (iface_ == I2C) {
    i2c_->beginTransmission(dev_);
    i2c_->write(reg);
    i2c_->endTransmission(false);
    bytes_rx_ = i2c_->requestFrom(static_cast<uint8_t>(dev_), count);
    if (bytes_rx_ == count) {
      for (size_t i = 0; i < count; i++) {
        data[i] = i2c_->read();
      }
      return true;
    } else {
      return false;
    }
  } else {
    spi_->beginTransaction(SPISettings(SPI_CLOCK_, MSBFIRST, SPI_MODE3));
    #if defined(TEENSYDUINO)
    digitalWriteFast(dev_, LOW);
    #else
    digitalWrite(dev_, LOW);
    #endif
    spi_->transfer(reg | SPI_READ_ | SPI_MS_);
    spi_->transfer(data, count);
    #if defined(TEENSYDUINO)
    digitalWriteFast(dev_, HIGH);
    #else
    digitalWrite(dev_, HIGH);
    #endif
    spi_->endTransaction();
    return true;
  }
}

}  // namespace bfs


examples/cmake/i2c.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2022 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "lis3mdl.h"

bfs::Lis3mdl mag(&Wire, bfs::Lis3mdl::I2C_ADDR_PRIM);

int main() {
  Serial.begin(115200);
  while (!Serial) {}
  Wire.begin();
  Wire.setClock(400000);
  if (!mag.Begin()) {
    Serial.println("Unable to communicate with LIS3MDL");
    while (1) {}
  }
  while (1) {
    if (mag.Read()) {
      Serial.print(mag.new_x_data());
      Serial.print("\t");
      Serial.print(mag.new_y_data());
      Serial.print("\t");
      Serial.print(mag.new_z_data());
      Serial.print("\t");
      Serial.print(mag.mag_x_ut());
      Serial.print("\t");
      Serial.print(mag.mag_y_ut());
      Serial.print("\t");
      Serial.print(mag.mag_z_ut());
      Serial.print("\n");
    }
  }
}


examples/cmake/spi.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2022 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "lis3mdl.h"

bfs::Lis3mdl mag(&SPI, 10);

bool status;
unsigned int t1, t2, t3, t4;

void drdy() {

}

int main() {
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("STARTING TEST");
  SPI.begin();
  bool status = mag.Begin();
  Serial.println(status);
  // attachInterrupt(9, drdy, RISING);
  while (1) {
    t3 = micros();
    status = mag.Read();
    t4 = micros();
    if (status) {
      t1 = micros();
      Serial.print(mag.new_x_data());
      Serial.print("\t");
      Serial.print(mag.new_y_data());
      Serial.print("\t");
      Serial.print(mag.new_z_data());
      Serial.print("\t");
      Serial.print(t1 - t2);
      Serial.print("\t");
      Serial.print(t4 - t3);
      Serial.print("\n");
      t2 = t1;
    }
  }
}


src/lis3mdl.h
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2022 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#ifndef LIS3MDL_SRC_LIS3MDL_H_  // NOLINT
#define LIS3MDL_SRC_LIS3MDL_H_

#if defined(ARDUINO)
#include <Arduino.h>
#include "Wire.h"
#include "SPI.h"
#else
#include <cstddef>
#include <cstdint>
#include "core/core.h"
#endif

namespace bfs {

class Lis3mdl {
 public:
  enum I2cAddr : uint8_t {
    I2C_ADDR_PRIM = 0x1C,
    I2C_ADDR_SEC = 0x1E
  };
  enum Range : uint8_t {
    RANGE_4GS =   0b00,
    RANGE_8GS =   0b01,
    RANGE_12GS =  0b10,
    RANGE_16GS =  0b11
  };
  enum Odr : uint8_t {
    ODR_0_625HZ = 0b0000,
    ODR_1_25HZ =  0b0010,
    ODR_2_5HZ =   0b0100,
    ODR_5HZ =     0b0110,
    ODR_10HZ =    0b1000,
    ODR_20HZ =    0b1010,
    ODR_40HZ =    0b1100,
    ODR_80HZ =    0b1110,
    ODR_155HZ =   0b0001,
    ODR_300HZ =   0b0011,
    ODR_560HZ =   0b0101,
    ODR_1000HZ =  0b0111
  };
  Lis3mdl() {}
  Lis3mdl(TwoWire *i2c, const I2cAddr addr) : i2c_(i2c),
                                              dev_(static_cast<uint8_t>(addr)),
                                              iface_(I2C) {}
  Lis3mdl(SPIClass *spi, const uint8_t cs) : spi_(spi), dev_(cs), iface_(SPI) {}
  void Config(TwoWire *i2c, const I2cAddr addr);
  void Config(SPIClass *spi, const uint8_t cs);
  bool Begin();
  bool ConfigOdr(const Odr odr);
  inline Odr odr() const {return odr_;}
  bool ConfigRange(const Range range);
  inline Range range() const {return range_;}
  bool Read();
  bool new_x_data();
  bool new_y_data();
  bool new_z_data();
  inline float mag_x_ut() const {return mag_ut_[0];}
  inline float mag_y_ut() const {return mag_ut_[1];}
  inline float mag_z_ut() const {return mag_ut_[2];}
  inline float die_temp_c() const {return temp_;}

 private:
  /* Comms */
  enum Interface : int8_t {
    SPI,
    I2C
  };
  TwoWire *i2c_;
  SPIClass *spi_;
  uint8_t dev_;
  Interface iface_;
  uint8_t bytes_rx_;
  static constexpr int32_t SPI_CLOCK_ = 7000000;
  static constexpr uint8_t SPI_READ_ = 0x80;
  static constexpr uint8_t SPI_MS_ = 0x40;
  static constexpr uint8_t LIS3MDL_WHOAMI_ = 0x3D;
  /* Config */
  enum PerfMode {
    PERF_MODE_LOW_POWER = 0b00,
    PERF_MODE_MEDIUM_POWER = 0b01,
    PERF_MODE_HIGH = 0b10,
    PERF_MODE_ULTRA_HIGH = 0b11
  };
  enum OpMode {
    OP_MODE_CONTINUOUS = 0b00,
    OP_MODE_SINGLE = 0b01,
    OP_MODE_POWERDOWN = 0b11
  };
  bool continuous_;
  uint8_t val_, data_, mask_;
  uint8_t who_am_i_;
  PerfMode perf_mode_;
  OpMode op_mode_;
  Odr odr_;
  Range range_;
  /* Data */
  uint8_t stat_;
  uint8_t buf_[9];
  int16_t x_, y_, z_, t_;
  bool data_updated_;
  bool new_x_data_, new_y_data_, new_z_data_;
  float mag_ut_[3];
  float temp_;
  float scale_;
  /* Register addresses */
  static constexpr uint8_t REG_WHO_AM_I_ = 0x0F;
  static constexpr uint8_t REG_CTRL_REG1_ = 0x20;
  static constexpr uint8_t REG_CTRL_REG2_ = 0x21;
  static constexpr uint8_t REG_CTRL_REG3_ = 0x22;
  static constexpr uint8_t REG_CTRL_REG4_ = 0x23;
  static constexpr uint8_t REG_CTRL_REG5_ = 0x24;
  static constexpr uint8_t REG_STATUS_REG_ = 0x27;
  static constexpr uint8_t REG_OUT_X_L_ = 0x28;
  static constexpr uint8_t REG_OUT_X_H_ = 0x29;
  static constexpr uint8_t REG_OUT_Y_L_ = 0x2A;
  static constexpr uint8_t REG_OUT_Y_H_ = 0x2B;
  static constexpr uint8_t REG_OUT_Z_L_ = 0x2C;
  static constexpr uint8_t REG_OUT_Z_H_ = 0x2D;
  static constexpr uint8_t REG_TEMP_OUT_L_ = 0x2E;
  static constexpr uint8_t REG_TEMP_OUT_H_ = 0x2F;
  static constexpr uint8_t REG_INT_CFG_ = 0x30;
  static constexpr uint8_t REG_INT_SRC_ = 0x31;
  static constexpr uint8_t REG_INT_THS_L_ = 0x32;
  static constexpr uint8_t REG_INT_THS_H_ = 0x33;
  /* Util functions */
  bool EnableBdu();
  bool EnableContinousMode(bool en);
  bool EnableTemp();
  bool SetPerfMode(const PerfMode perf_mode);
  bool SetOpMode(const OpMode op_mode);
  bool SetDataRate(const Odr data_rate);
  bool SetRange(const Range range);
  bool GetPerfMode(PerfMode * const perf_mode);
  bool GetOpMode(OpMode * const op_mode);
  bool GetDataRate(Odr * const data_rate);
  bool GetRange(Range * const range);
  bool Read(const uint8_t reg, const uint8_t bits, const uint8_t shift,
            uint8_t * const data);
  bool Write(const uint8_t reg, const uint8_t data, const uint8_t bits,
             const uint8_t shift);
  bool WriteRegister(const uint8_t reg, const uint8_t data);
  bool ReadRegisters(const uint8_t reg, const uint8_t count,
                     uint8_t * const data);
};

}  // namespace bfs

#endif  // LIS3MDL_SRC_LIS3MDL_H_ NOLINT
